### PR TITLE
New slab calendar

### PR DESCRIPTION
**Pull request by tmihordea**
New Calendar UI with Charts
 - replace calendar UI with one that has supports multi-view (daily, monthly, yearly). Each view has renderers for its elements and they are used to display the SLO accordingly
 - include an additional view that shows all the SLO data as a bar chart, with a brush control for scroll and zoom
- updated the api for /status to return the hourly slo for the last 2 years

### CODE REVIEW

File: src/main/scala/com/criteo/slab/core/Board.scala [hunk 1]
```diff
@@ -17,7 +17,8 @@ case class Board[B <: HList](
                               boxes: B,
                               aggregate: (Map[Box[_], View], Context) => View,
                               layout: Layout,
-                              links: Seq[(Box[_], Box[_])] = Seq.empty
+                              links: Seq[(Box[_], Box[_])] = Seq.empty,
+                              slo: Double = 0.97
```
Comment (jedirandy) [inline src/main/scala/com/criteo/slab/core/Board.scala]: can you update the comment to describe this argument?

File: src/main/scala/com/criteo/slab/app/StateService.scala [hunk 2]
```diff
@@ -56,16 +56,46 @@ private[slab] class StateService(
...
+          val allResults =
+            for {idx <- -1 to months + 1} yield {
+              val from = fromStartOfMonth.plus(idx, ChronoUnit.MONTHS)
+              val to = fromStartOfMonth.plus(idx + 1, ChronoUnit.MONTHS)
+              if (from.compareTo(now) < 0)
+                boardMonthlySloInner(board, from, to)
+              else
+                Future.successful(Map.empty[Long, Double])
+            }
+          val fromTs = from.toEpochSecond * 1000
```
Comment (jedirandy) [inline src/main/scala/com/criteo/slab/app/StateService.scala]: you can use toEpochMilli

File: src/main/scala/com/criteo/slab/app/StateService.scala [hunk 3]
```diff
@@ -56,16 +56,46 @@ private[slab] class StateService(
...
+            executor.fetchHourlySlo(from.toInstant, until.toInstant).map(_.toMap)
+          }
+
+          val from = now.minus(statsDays, ChronoUnit.DAYS)
+          val until = now
+
+          val fromStartOfMonth = from.truncatedTo(ChronoUnit.DAYS).withDayOfMonth(1)
+
+          val months = ChronoUnit.MONTHS.between(fromStartOfMonth, until).toInt
+          //take take one extra period to avoid timezone issues
```
Comment (jedirandy) [inline src/main/scala/com/criteo/slab/app/StateService.scala]: do we really need extra months to handle this? I guess an offset by day would cover all zones

File: src/main/scala/com/criteo/slab/lib/graphite/GraphiteStore.scala [hunk 4]
```diff
@@ -20,22 +21,23 @@ import scala.util.{Failure, Success, Try}
...
+  * @param maxConnections Max connections of the Http client
+  * @param ec             The execution context
   */
 class GraphiteStore(
                      host: String,
                      port: Int,
                      webHost: String,
                      checkInterval: Duration,
                      group: Option[String] = None,
+                     sloGroup: Option[String] = None,
```
Comment (jedirandy) [inline src/main/scala/com/criteo/slab/lib/graphite/GraphiteStore.scala]: update the comment too

File: src/main/webapp/js/components/Calendar.js [hunk 5]
```diff
@@ -1,77 +1,227 @@
...
+                        <XAxis dataKey="date" tickFormatter={this.xAxisTickFormatter}/>
+                        <YAxis domain={buildDomain(data, slo)}
+                               tickFormatter={this.yAxisTickFormatter}/>
+                        <Tooltip content={<CustomChartTooltip payload label active/>}/>
+                        <Brush dataKey='date' height={30} tickFormatter={this.xAxisTickFormatter}
+                               startIndex={Math.max(0, data.length - 31)}
+                               endIndex={Math.max(0, data.length - 1)}
+                               travellerWidth={10}/>
+                        <Bar dataKey="percent" shape={<ThresholdFillBar y x width height value threshold={slo}/>}/>
+                        <ReferenceLine y={slo} label="SLO"/>
```
Comment (jedirandy) [inline src/main/webapp/js/components/Calendar.js]: can we put the "SLO" title on the top of the box in white color?

File: src/main/webapp/js/components/Calendar.js [hunk 6]
```diff
@@ -1,77 +1,227 @@
...
+                        renderMonth={this.renderMonth(stats && stats.monthly, slo)}
+                        renderYear={this.renderYear(stats && stats.yearly, slo)}
+                        onChange={onDayClick}
+                        isValidDate={this.validateDate}
+              />
+              { data && data.length > 0 &&
+                (<ResponsiveContainer className="back" width="100%" height="100%">
+                    <BarChart data={data} margin={{left: -20}}>
+                        <XAxis dataKey="date" tickFormatter={this.xAxisTickFormatter}/>
+                        <YAxis domain={buildDomain(data, slo)}
```
Comment (jedirandy) [inline src/main/webapp/js/components/Calendar.js]: maybe we can set it to be always 0-100% for consistency?

File: src/main/webapp/js/components/Calendar.js [hunk 7]
```diff
@@ -1,77 +1,227 @@
...
+                               tickFormatter={this.yAxisTickFormatter}/>
+                        <Tooltip content={<CustomChartTooltip payload label active/>}/>
+                        <Brush dataKey='date' height={30} tickFormatter={this.xAxisTickFormatter}
```
Comment (jedirandy) [inline src/main/webapp/js/components/Calendar.js]: the date texts are hidden when close to the ends, a fix on this?

File: webpack.config.js [hunk 8]
```diff
@@ -30,10 +30,11 @@ var config = {
           presets: ['react', 'es2015'],
           plugins: [
             'transform-class-properties',
-            'transform-object-rest-spread'
+            'transform-object-rest-spread',
+            'transform-exponentiation-operator'
           ]
         },
-        include: path.join(__dirname, 'src')
+        include: [path.join(__dirname, 'src'), path.resolve(__dirname, "node_modules/recharts/es6")] //recharts es6 needed
```
Comment (jedirandy) [inline webpack.config.js]: why is it necessary to add this resolve path?

File: src/main/scala/com/criteo/slab/app/StateService.scala (hunk 2 above)
Comment (tmihordea) [reply src/main/scala/com/criteo/slab/app/StateService.scala]: ZonedDateTime doesn't have toEpochMilli

File: src/main/webapp/js/components/Calendar.js (hunk 6 above)
Comment (crotundu) [reply src/main/webapp/js/components/Calendar.js]: Forcing it to 0-100% will strip away the amplitude detail if your bars are close to each other in value and in our case most of the times we're in the 90-100% range.

File: src/main/webapp/js/components/Calendar.js (hunk 7 above)
Comment (crotundu) [reply src/main/webapp/js/components/Calendar.js]: I will make the brush shorter to accommodate the ends.

File: src/main/webapp/js/components/Calendar.js (hunk 5 above)
Comment (crotundu) [reply src/main/webapp/js/components/Calendar.js]: The idea was to add context to the red line itself.

File: webpack.config.js (hunk 8 above)
Comment (crotundu) [reply webpack.config.js]: It was meant as a temporary workaround for some issues in recharts sources that have since been fixed. I will remove it.

**Pull request merged**